File: TimberCruisingApp/Persistence/Repositories/SpeciesConfigRepository.swift
// Spec §8 + REQ-PRJ-005.

import Foundation
import CoreData
import Models

public protocol SpeciesConfigRepository {
    func create(_ s: SpeciesConfig) throws -> SpeciesConfig
    func read(code: String) throws -> SpeciesConfig?
    func update(_ s: SpeciesConfig) throws -> SpeciesConfig
    func delete(code: String) throws
    func list() throws -> [SpeciesConfig]
}

public final class CoreDataSpeciesConfigRepository: SpeciesConfigRepository {
    private let stack: CoreDataStack
    public init(stack: CoreDataStack) { self.stack = stack }

    public func create(_ s: SpeciesConfig) throws -> SpeciesConfig {
        try performWrite(stack: stack) { ctx in
            let e = try insert(SpeciesConfigEntity.self, entityName: "SpeciesConfigEntity", in: ctx)
            SpeciesConfigMapper.apply(s, to: e)
            return s
        }
    }

    public func read(code: String) throws -> SpeciesConfig? {
        try performRead(stack: stack) { ctx in
            guard let e = try fetchOne(SpeciesConfigEntity.self, entityName: "SpeciesConfigEntity",
                                       keyPath: "code", value: code as CVarArg, in: ctx)
            else { return nil }
            return SpeciesConfigMapper.toStruct(e)
        }
    }

    public func update(_ s: SpeciesConfig) throws -> SpeciesConfig {
        try performWrite(stack: stack) { ctx in
            guard let e = try fetchOne(SpeciesConfigEntity.self, entityName: "SpeciesConfigEntity",
                                       keyPath: "code", value: s.code as CVarArg, in: ctx)
            else { throw CoreDataError.notFound(id: s.code) }
            SpeciesConfigMapper.apply(s, to: e)
            return s
        }
    }

    public func delete(code: String) throws {
        try performWrite(stack: stack) { ctx in
            guard let e = try fetchOne(SpeciesConfigEntity.self, entityName: "SpeciesConfigEntity",
                                       keyPath: "code", value: code as CVarArg, in: ctx)
            else { throw CoreDataError.notFound(id: code) }
            ctx.delete(e)
        }
    }

    public func list() throws -> [SpeciesConfig] {
        try performRead(stack: stack) { ctx in
            let sort = [NSSortDescriptor(key: "code", ascending: true)]
            return try fetchMany(SpeciesConfigEntity.self, entityName: "SpeciesConfigEntity",
                                 sort: sort, in: ctx)
                .map(SpeciesConfigMapper.toStruct)
        }
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/data/cruise/SpeciesConfigRepository.kt
// Port of iOS Persistence/Repositories/SpeciesConfigRepository.swift.
// Spec §8 + REQ-PRJ-005.

package com.hcjeong.forestix.data.cruise

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

interface SpeciesConfigRepository {
    suspend fun create(s: SpeciesConfig): SpeciesConfig
    suspend fun read(code: String): SpeciesConfig?
    suspend fun update(s: SpeciesConfig): SpeciesConfig
    suspend fun delete(code: String)
    suspend fun list(): List<SpeciesConfig>

    /// Live view of `list()` (code ascending) for Compose screens.
    fun observeList(): Flow<List<SpeciesConfig>>
}

class RoomSpeciesConfigRepository(private val dao: SpeciesConfigDao) : SpeciesConfigRepository {

    override suspend fun create(s: SpeciesConfig): SpeciesConfig {
        dao.upsert(SpeciesConfigMapper.apply(s))
        return s
    }

    override suspend fun read(code: String): SpeciesConfig? =
        dao.byCode(code)?.let(SpeciesConfigMapper::toStruct)

    override suspend fun update(s: SpeciesConfig): SpeciesConfig {
        dao.byCode(s.code) ?: throw CruiseDataError.NotFound(s.code)
        dao.upsert(SpeciesConfigMapper.apply(s))
        return s
    }

    override suspend fun delete(code: String) {
        if (dao.deleteByCode(code) == 0) throw CruiseDataError.NotFound(code)
    }

    override suspend fun list(): List<SpeciesConfig> =
        dao.list().map(SpeciesConfigMapper::toStruct)

    override fun observeList(): Flow<List<SpeciesConfig>> =
        dao.observeList().map { rows -> rows.map(SpeciesConfigMapper::toStruct) }
}
